hedera-node/hedera-app/src/main/java/com/hedera/node/app/util/HashTreeManager.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.util;
+
+import com.hedera.pbj.runtime.Codec;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.util.ArrayList;
+import java.util.List;
+
+public class HashTreeManager<T> {
+
+    public HashTreeManager(Codec<T> codec) {
+        this.codec = codec;
+    }
+
+    private final Codec<T> codec;
+    List<Bytes> inputsCompleteSubtreeHashes;
+    List<Bytes> outputsCompleteSubtreeHashes;
+    private List<Bytes> hashList = new ArrayList<>();
+
+    private enum Alternator {
+        LEFT,
+        RIGHT
+    }
+
+    Alternator currentAlternate = Alternator.LEFT;
+
+    public void processNodes(List<T> elements) {
+        for (T element : elements) {
+            Bytes elementHash = hash(codec.toBytes(element));
+            hashList.add(elementHash);
+
+            if ((elements.indexOf(element) & 1) == 1) {
+                Bytes yHash = hashList.remove(hashList.size() - 1);
+                Bytes xHash = hashList.remove(hashList.size() - 1);
+                Bytes combinedHash = hash(yHash.append(xHash));
+                hashList.add(combinedHash);
+            }
+        }
```

Comment:
This logic exists in various separate places in this class. Intentionally not consolidated?

Reply:
Refactored now 👍